library/include/DataFuncs.h

```diff
@@ -34,229 +34,106 @@ distribution.
+        typedef rT type;
+        typedef CT class_type;
+        static const bool is_method = true;
+    };
 
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6, class A7, class A8
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6,A7,A8))
-INSTANTIATE_WRAPPERS(8, (A1,A2,A3,A4,A5,A6,A7,A8), (OSTREAM_ARG,A1,A2,A3,A4,A5,A6,A7,A8),
-                        (vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8), (out,vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6); LOAD_ARG(A7); LOAD_ARG(A8);)
-#undef FW_TARGS
+/*
+ * This is required to make GCC accept std::tuple{ (DFHack::Lua::Get<aT>(L,base++))... };
+ * The order of effects is guaranteed since C+11
+ * See: https://en.cppreference.com/w/cpp/language/eval_order (point 9)
+ */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsequence-point"
+
+#define CALL_AND_PUSH(rT,fun,args) \
+    if constexpr (std::is_same_v<rT,void>){ \
+        std::apply(fun,args); \
+    } else { \
+        rT rv = std::apply(fun,args); \
+        df::identity_traits<rT>::get()->lua_read(L, UPVAL_METHOD_NAME, &rv); \
+    }
 
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6, class A7, class A8, class A9
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6,A7,A8,A9))
-INSTANTIATE_WRAPPERS(9, (A1,A2,A3,A4,A5,A6,A7,A8,A9),
-                        (OSTREAM_ARG,A1,A2,A3,A4,A5,A6,A7,A8,A9),
-                        (vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8,vA9),
-                        (out,vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8,vA9),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6); LOAD_ARG(A7); LOAD_ARG(A8);
-                     LOAD_ARG(A9);)
-#undef FW_TARGS
+    template<typename T> struct function_wrapper {};
 
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6, class A7, class A8, class A9, class A10
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6,A7,A8,A9,A10))
-INSTANTIATE_WRAPPERS(10, (A1,A2,A3,A4,A5,A6,A7,A8,A9,A10),
-                         (OSTREAM_ARG,A1,A2,A3,A4,A5,A6,A7,A8,A9,A10),
-                         (vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8,vA9,vA10),
-                         (out,vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8,vA9,vA10),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6); LOAD_ARG(A7); LOAD_ARG(A8);
-                     LOAD_ARG(A9); LOAD_ARG(A10);)
-#undef FW_TARGS
+    template<typename rT, typename ...aT>
+    struct function_wrapper<rT(*)(DFHack::color_ostream&, aT...)> {
+        static const int num_args = sizeof...(aT);
+        static void execute(lua_State *L, int base, rT (fun)(DFHack::color_ostream& out, aT...)) {
+             cur_lua_ostream_argument out(L);
+             auto args = std::tuple<DFHack::color_ostream&, aT...>{ out, (DFHack::Lua::Get<aT>(L,base++))... };
```

Comment:
it would make the code much uglier. the sequence point issue is _solely_ due to paternalism by the gcc team, choosing to implement a warning for code that would be UB in C++03, except that it's not syntactically valid in C++03 anyway so why even bother?

i'm totally on board with the use of the pragma here

Reply:
that said, i've believe i found a solution that avoids the issue by using `index_sequence_for`